### PR TITLE
Extend YAML load functionality to *LIST and multi-resources

### LINKED ISSUE #671: Extend #655 with a List Parser
Recently I proposed #655 that will work with Objects and Simple kinds. (See https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md) It is however, not capable of taking a list yet.  I'll work on a kind: List parser that can create multiple objects at a time.
/assign

### PR DESCRIPTION
**Pull request by micw523**
Fixes #671 
This PR addresses #504 #671, which is an extension of #655.

It allows direct loading of YAMLs such as in
https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/multi-resource-yaml.yaml
and
https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/list.yaml

Important changes:
- The function will no longer return an API object after processing the YAML file due to concerns of partial / complete failure during the creation process.

### CODE REVIEW

**Comment by micw523** [comment]
/unhold
Merge conflicts solved

File: kubernetes/utils/create_from_yaml.py [hunk 1]
```diff
@@ -23,51 +23,85 @@
...
+                for yml_object in yml_document["items"]:
+                    k8s_api_all.append(create_from_yaml_single_item(
+                        k8s_client, yml_object, verbose, **kwargs))
+            else:
+                # This is a single object. Call the single item method
+                k8s_api_all.append(create_from_yaml_single_item(
+                    k8s_client, yml_document, verbose, **kwargs))
+    if output_list is False:
+        if len(k8s_api_all) == 1:
+            return k8s_api_all[0]
```
Comment (roycaihw) [inline kubernetes/utils/create_from_yaml.py]: Should this be returning the first item when `output_list is False` or `len(k8s_api_all) == 1`?

File: kubernetes/utils/create_from_yaml.py [hunk 2]
```diff
@@ -23,51 +23,85 @@
...
     """
     Perform an action from a yaml file. Pass True for verbose to
     print confirmation information.
     Input:
     yaml_file: string. Contains the path to yaml file.
     k8s_cline: an ApiClient object, initialized with the client args.
+    verbose: If True, print confirmation from the create action. Default is False.
+    output_list: compatibility option with v8.0.0. Default is False.
+        Function returns a single api object when there is only one when set False.
+        Does not affect when multiple objects are generated.
```
Comment (roycaihw) [inline kubernetes/utils/create_from_yaml.py]: I don't think it's a breaking change, because we didn't support list creation. 

I prefer not having the `output_list` parameter. Clients that create individual resources won't see difference, right?

File: kubernetes/utils/create_from_yaml.py [hunk 3]
```diff
@@ -23,51 +23,85 @@
...
+                    k8s_api_all.append(create_from_yaml_single_item(
+                        k8s_client, yml_object, verbose, **kwargs))
+            else:
+                # This is a single object. Call the single item method
+                k8s_api_all.append(create_from_yaml_single_item(
+                    k8s_client, yml_document, verbose, **kwargs))
+    if output_list is False:
+        if len(k8s_api_all) == 1:
+            return k8s_api_all[0]
+    return k8s_api_all
```
Comment (roycaihw) [inline kubernetes/utils/create_from_yaml.py]: Have you verified kubectl's behavior? What if we failed creating some of the items?

File: kubernetes/utils/create_from_yaml.py (hunk 2 above)
Comment (micw523) [reply kubernetes/utils/create_from_yaml.py]: I think only if we have this right now the clients won’t see any difference. If we output lists all the time, we would have k8s_api returned in the current release and [k8s_api] with this PR if we are creating individual resource. I’m thinking to let the client choose what they want to do: if they want to create everything without checking the return type they can set ```output_list``` to ```True``` and just iterate the list. Otherwise if they want the legacy behavior they can choose not to pass any parameter.

File: kubernetes/utils/create_from_yaml.py (hunk 1 above)
Comment (micw523) [reply kubernetes/utils/create_from_yaml.py]: See above - I’m think about the case when the user do not want to check the return type and just want to iterate a list.

File: kubernetes/utils/create_from_yaml.py (hunk 3 above)
Comment (micw523) [reply kubernetes/utils/create_from_yaml.py]: I think I only tested on 404 with an invalid API version and 409 with conflicts. These exceptions are handled by API internally.

I probably should go test with a multi resource yaml with one of the resources being 404/409. I’ll keep you updated.

**Comment by micw523** [comment]
The base module update is an overlook. I should fix that...

**Comment by micw523** [comment]
/unshrug
Maybe should send you a ticket to get these categories @krzyzacy

**Comment by krzyzacy** [comment]
noooooooooooo when psychedelic stopped work?! cc @fejta :joy:

/meow caturday

**Comment by fejta** [comment]
> noooooooooooo when psychedelic stopped work?!

Dunno... We just send it to thecatapi, not sure why it's including it in the list but not returning a valid image. https://api.thecatapi.com/api/images/get?format=json&results_per_page=1&category=psychedelic returns an empty list is the problem

**Comment by micw523** [comment]
@fetja I still want my code reviewed here😂
On a side note, I replied in
kubernetes/test-infra#10111 
Available categories are 5, 6, 15, 9, 3, 1, 10, 14, 2, 4, 7 from curl. The api list is out of date. No boys, no psychedelic, no girls, no breaded.
/meowvie sunglasses

File: kubernetes/utils/create_from_yaml.py [hunk 4]
```diff
@@ -23,51 +23,105 @@
...
+
+    Available parameters for creating <kind>:
     :param async_req bool
     :param bool include_uninitialized: If true, partially initialized resources are included in the response.
     :param str pretty: If 'true', then the output is pretty printed.
     :param str dry_run: When present, indicates that modifications should not be persisted. An invalid or unrecognized dryRun directive will result in an error response and no further processing of the request. Valid values are: - All: all dry run stages will be processed
     """
 
+    k8s_api_all = []
     with open(path.abspath(yaml_file)) as f:
```
Comment (amih90) [inline kubernetes/utils/create_from_yaml.py]: `path.abspath(yaml_file) `
will crash on string input while `yaml.load(str)` can load the yaml from a string.
very useful if you need to change the yaml dynamically

ill suggest to support yaml_file as string and a file

Comment (micw523) [reply kubernetes/utils/create_from_yaml.py]: I don’t think this is necessary. The feature requested is to replicate kubectl create -f which of course would not support strings. YAML files can be load as a dictionary and passed as the body parameter when the user calls the corresponding APIs. I don’t see the benefit of changing a YAML on the fly - we’re not doing kubectl apply -f here.

If you want this feature to parse as a string, please file a separate issue.

File: kubernetes/utils/create_from_yaml.py [hunk 5]
```diff
@@ -23,51 +23,105 @@
...
+                            "Error when creating {0}/{1}".format(
+                                yml_object["kind"],
+                                yml_object["metadata"]["name"]),
+                            api_exception)
+            else:
+                # This is a single object. Call the single item method
+                try:
+                    k8s_api_all.append(create_from_yaml_single_item(
+                        k8s_client, yml_document, verbose, **kwargs))
+                except client.rest.ApiException as api_exception:
```
Comment (roycaihw) [inline kubernetes/utils/create_from_yaml.py]: it is great that the util runs through the entire list even when there is a failure, like kubectl does

however kubectl returns error code 1 to stderr when failure happens, I'd expect we raise error to upstream after we finish the entire list, to make it easier for automation

File: kubernetes/e2e_test/test_utils.py [hunk 6]
```diff
@@ -112,5 +112,85 @@ def test_api_service(self):
...
+        self.assertEqual("v1", ctr_api.get_api_resources().group_version)
+        ctr = ctr_api.read_namespaced_replication_controller(
+            name="mock", namespace="default")
+        self.assertIsNotNone(ctr)
+        ctr_api.delete_namespaced_replication_controller(name="mock",
+            namespace="default", body={})
+        svc_api.delete_namespaced_service(name="mock",
+            namespace="default", body={})
+
+    def test_multi_resource_with_conflict(self):
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: need to document the tests. it's hard to follow what's expected from reading the test and file names

File: kubernetes/utils/create_from_yaml.py (hunk 5 above)
Comment (roycaihw) [reply kubernetes/utils/create_from_yaml.py]: instead of `print` each failure. I'd like to see we concatenate the errors and raise in the end

for question about what the raised error should look like, as long as the error clearly identifies failed objects and is well-formatted to parse (for upstream automation) when necessary, I'm okay with having either a string or a new class for the error message. @mbohlool @yliaog Do you have suggestion on what's the best practice?

below is what kubectl does
```
$ kubectl create -f hack/testdata/list.yaml 1>/dev/null
Error from server (AlreadyExists): services "list-service-test" already exists
Error from server (AlreadyExists): deployments.apps "list-deployment-test" already exists
```

File: kubernetes/e2e_test/test_utils.py [hunk 7]
```diff
@@ -17,100 +17,282 @@
...
     @classmethod
     def setUpClass(cls):
         cls.config = base.get_e2e_configuration()
 
     def test_app_yaml(self):
+        """
+        Test for creating and deleting a deployment on default namespace.
+        Using apps API
+        Using yaml file apps-deployment.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: suggest
```
    def test_create_apps_deployment_from_yaml(self):
        """
        Should be able to create a apps/v1beta1 deployment.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 8]
```diff
@@ -17,100 +17,282 @@
...
+            name="nginx-app", namespace="default",
             body={})
-        
+
     def test_extension_yaml(self):
+        """
+        Test for creating and deleting a deployment on default namespace.
+        Using extensions API
+        Using yaml file extensions-deployment.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: suggest
```
    def test_create_extensions_deployment_from_yaml(self):
        """
        Should be able to create an extensions/v1beta1 deployment.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 9]
```diff
@@ -17,100 +17,282 @@
...
+            name="nginx-deployment", namespace="default",
             body={})
-    
+
     def test_core_pod_yaml(self):
+        """
+        Test for creating and deleting a pod on default namespace.
+        Using core API
+        Using yaml file core-pod.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: suggest
```
    def test_create_pod_from_yaml(self):
        """
        Should be able to create a pod.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 10]
```diff
@@ -17,100 +17,282 @@
...
+        k8s_api.delete_namespaced_pod(
             name="myapp-pod", namespace="default",
             body={})
 
     def test_core_service_yaml(self):
+        """
+        Test for creating and deleting a service on default namespace.
+        Using core API
+        Using yaml file core-service.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]:
```
    def test_create_service_from_yaml(self):
        """
        Should be able to create a service.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 11]
```diff
@@ -17,100 +17,282 @@
...
             name="my-service", namespace="default",
             body={})
-        
+
     def test_core_namespace_yaml(self):
+        """
+        Test for creating and deleting namespace
+        Using core API
+        Using yaml file core-namespace.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: def test_create_namespace_from_yaml(self):
        """
        Should be able to create a namespace.
        """

File: kubernetes/e2e_test/test_utils.py [hunk 12]
```diff
@@ -17,100 +17,282 @@
...
 
     def test_deployment_in_namespace(self):
+        """
+        Test: Create a namespace. Create and delete a deployment on
+        the namespace just created. Delete the namespace created at the
+        beginning of the test.
+        Using core, extensions API
+        Using yaml file core-namespace-dep.yaml
+        Using yaml file extensions-deployment-dep.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]:
```
def test_create_deployment_in_namespace_from_yaml(self):
    """
    Should be able to create a deployment in a non-default namespace.
    """
```

File: kubernetes/e2e_test/test_utils.py [hunk 13]
```diff
@@ -17,100 +17,282 @@
...
+        core_api.delete_namespace(name="dep", body={})
+
+    def test_api_service_with_conflict(self):
+        """
+        Test: Create an APIService. Recreate the APIService to test conflict
+        handling. Delete the APIService.
+        Check the program raises CreationFailedError.
+        Using apiregistration API
+        Using yaml file api-service.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]:
```
    def test_create_apiservice_from_yaml_with_conflict(self):
        """
        Creating the same apiservice twice should fail due to conflict.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 14]
```diff
@@ -17,100 +17,282 @@
...
+            k8s_client, "kubernetes/e2e_test/test_yaml/api-service.yaml")
+        self.assertEqual("apiregistration.k8s.io/v1beta1",
+                         k8s_api.get_api_resources().group_version)
         svc = k8s_api.read_api_service(
             name="v1alpha1.wardle.k8s.io")
         self.assertIsNotNone(svc)
-        resp = k8s_api.delete_api_service(
-            name="v1alpha1.wardle.k8s.io", body={})
\ No newline at end of file
+        with self.assertRaises(utils.FailToCreateError):
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: could you also verify the error message from apiserver?

File: kubernetes/e2e_test/test_utils.py [hunk 15]
```diff
@@ -17,100 +17,282 @@
...
+        k8s_api.delete_api_service(
+            name="v1alpha1.wardle.k8s.io", body={})
+
+    def test_list(self):
+        """
+        Test for creating and deleting a service and a deployment using
+        the List kind.
+        Using core, extensions API
+        Using yaml file list.yaml
+        """
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: suggest
```
    def test_create_from_yaml_list(self):
        """
        Should be able to create multiple objects from a yaml of list kind.
        """
```

File: kubernetes/e2e_test/test_utils.py [hunk 16]
```diff
@@ -17,100 +17,282 @@
...
+        k8s_client = client.api_client.ApiClient(configuration=self.config)
+        k8s_api = utils.create_from_yaml(
+            k8s_client, "kubernetes/e2e_test/test_yaml/list.yaml")
+        svc_api = k8s_api[0]
+        self.assertEqual("v1", svc_api.get_api_resources().group_version)
+        svc = svc_api.read_namespaced_service(name="list-service-test",
+                                              namespace="default")
+        self.assertIsNotNone(svc)
+        ext_api = k8s_api[1]
```
Comment (roycaihw) [inline kubernetes/e2e_test/test_utils.py]: I don't see the point of doing so. It's typed client anyway

it's not a good pattern to iterate through a list of arbitrary api objects and call typed api functions imo

**Comment by thiagosantosleite** [comment]
Any plans to delete and apply?

kubectl delete -f <yaml file>
kubectl apply -f <yaml file>
